### PR TITLE
DDF-6166 Refactor Security Features File

### PR DESCRIPTION
**Pull request by bakejeyner**
Rename platform-session-invalidator to platform-session-invalidator-api
Rename security-logout-service to security-servlet-logout-api
Rename security-servlet-web-socket to security-servlet-web-socket-api
Rename session-management-service to session-management-api
Refactor security features

#### What does this PR do?
Simplifies the security feature file. This change will make the feature file easier to understand and maintain.

#### Who is reviewing it? 
@blen-desta 
@garrettfreibott 

#### How should this be tested?
[See the tests steps on the Alliance branch](https://github.com/codice/alliance/pull/829).

#### What are the relevant tickets?
Fixes: #6166

#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Threat Dragon models
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

File: features/security/src/main/feature/feature.xml [hunk 1]
```diff
@@ -21,694 +21,317 @@
...
-    <feature name="platform-http-proxy" version="${project.version}"
-             description="HTTP to HTTPS proxy.">
-        <feature>camel-jetty9</feature>
-        <feature>security-core-api</feature>
-        <feature>camel-blueprint</feature>
-        <feature>camel-http</feature>
-        <bundle>mvn:org.codice.thirdparty/commons-httpclient/3.1.0_1</bundle>
-        <bundle>mvn:ddf.platform/platform-http-proxy/${project.version}</bundle>
-    </feature>
+        <!--  TODO (https://github.com/codice/ddf/issues/5327)
```
Comment (blen-desta) [inline features/security/src/main/feature/feature.xml:296]: ❓ This issue is shown as closed in github. Should it be re-opened?

File: features/security/src/main/feature/feature.xml [hunk 2]
```diff
@@ -21,694 +21,317 @@
...
+        <bundle>mvn:ddf.platform.email/platform-email-impl/${project.version}</bundle>
 
-    <feature name="security-command-listener" version="${project.version}"
-             description="Audits all commands entered on the Karaf command line.">
-        <feature>security-core-api</feature>
-        <feature>security-core-impl</feature>
-        <feature>security-core-services</feature>
+        <!-- Miscellaneous -->
+        <bundle>mvn:ddf.platform/platform-paxweb-jettyconfig/${project.version}</bundle>
+        <bundle>mvn:ddf.platform/platform-http-proxy/${project.version}</bundle>
```
Comment (SmithJosh) [inline features/security/src/main/feature/feature.xml:277]: Was the http proxy enabled by default before? Imo it might be better to leave it uninstalled out of the box, and let people install it if they need it.

File: features/security/src/main/feature/feature.xml [hunk 3]
```diff
@@ -21,694 +21,317 @@
...
-        <bundle>wrap:mvn:com.nimbusds/nimbus-jose-jwt/${nimbus.version}</bundle>
-        <bundle>wrap:mvn:com.nimbusds/oauth2-oidc-sdk/${nimbus.oidc.version}</bundle>
-        <bundle>mvn:com.nimbusds/content-type/${nimbus.content-type.version}</bundle>
-        <bundle>mvn:com.nimbusds/lang-tag/${nimbus.langtag.version}</bundle>
-        <bundle>wrap:mvn:com.github.stephenc.jcip/jcip-annotations/1.0-1</bundle>
-        <bundle>mvn:commons-codec/commons-codec/${commons-codec.version}</bundle>
-        <bundle>mvn:ddf.platform.security/security-oidc-bundle/${project.version}</bundle>
+        <!-- Miscellaneous -->
+        <bundle>mvn:com.google.code.gson/gson/${gson.version}</bundle>
+        <bundle>wrap:mvn:org.springframework.ldap/spring-ldap-core/2.3.2.RELEASE</bundle>
```
Comment (SmithJosh) [inline features/security/src/main/feature/feature.xml:189]: Should all ldap functionality be contained in the ldap-auth feature? (and if this isn't used for ldap auth, I'm curious what it is used for)

File: features/security/src/main/feature/feature.xml (hunk 1 above)
Comment (bakejeyner) [reply features/security/src/main/feature/feature.xml:297]: It's still an issue, but I'm not sure when we are planning to get around to resolving it. I'm not sure what the best practice is in this case.

File: features/security/src/main/feature/feature.xml [hunk 4]
```diff
@@ -21,694 +21,317 @@
...
+        </configfile>
+        <bundle>mvn:ddf.security.claims/security-claims-ldap/${project.version}</bundle>
     </feature>
 
-    <feature name="security-services-app" version="${project.version}"
-             description="The Security Application provides Authentication, Authorization, and Auditing services.\nThey comprise both a framework that developers and integrators can extend and also a reference implementation that can be used which meets security requirements.::Security">
-        <feature>security-core</feature>
-        <feature>security-certificate</feature>
-        <feature>security-servlet-logout</feature>
+    <feature name="embedded-ldap" version="${project.version}"
```
[the remark this answers is not in the document]
Comment (bakejeyner) [reply features/security/src/main/feature/feature.xml:325]: The old `embedded-ldap` features only contained config files, so yes.

However this comment just reminded me that I haven't tried running itests yet 😱

File: features/apps/src/main/feature/feature.xml [hunk 5]
```diff
@@ -63,13 +62,17 @@
     <feature name="ddf-boot-features" version="${project.version}"
              description="Features that will be started on container startup.">
         <feature>ddf-core</feature>
-        <feature>security-services-app</feature>
+        <feature>security-core-services</feature>
+        <feature>oidc-auth</feature>
+        <feature>saml-auth</feature>
+        <feature>ldap-auth</feature>
```
[the remark this answers is not in the document]
Comment (bakejeyner) [reply features/apps/src/main/feature/feature.xml:68]: Good catch, made the change.

File: features/security/src/main/feature/feature.xml [hunk 6]
```diff
@@ -21,694 +21,317 @@
...
-             description="API for authentication handlers for web applications.">
-        <feature>security-core-api</feature>
-        <feature>parser-xml</feature>
-        <feature>oidc-dependencies</feature>
-        <bundle>mvn:org.apache.cxf/cxf-rt-rs-security-jose/${cxf.version}</bundle>
-        <bundle>mvn:org.apache.cxf/cxf-rt-rs-security-jose-jaxrs/${cxf.version}</bundle>
-        <bundle>mvn:ddf.security.handler/security-handler-api/${project.version}</bundle>
-    </feature>
+        <!-- Email -->
+        <bundle>mvn:ddf.platform.email/platform-email-api/${project.version}</bundle>
```
[the remark this answers is not in the document]
Comment (bakejeyner) [reply features/security/src/main/feature/feature.xml:101]: Oman this blew up.

I too thought it was weird that there were platform packages in here. They do depend on security packages, but I decided moved them over to the `utilities` feature file anyway.

But when I did that, the `utilities` feature itests they claimed they couldn't find the `org.codice.ddf.platform.util` package, even though I added the `platform-util` feature which exposes that package. No matter what I did, the email bundles just were not picking up that package.

The only way I could fix it was by including the big `security-core-services` feature. I have a feeling its because in `platform-util`'s bundle manifest the `org.codice.ddf.platform.util` exported package "uses" a ton of other packages that `security-core-services` ends up pulling in.

I couldn't find an easy way to determine those packages, so I'm just going to make the platform bundles depend on `security-core-services`. I have verified that we only pull in those platform bundles when we have already pulled in `security-core-services` anyway. I don't like it, but I also don't feel like doing an insane amount of trial and error or reading up on package resolution in karaf.

File: features/security/src/main/feature/feature.xml (hunk 2 above)
Comment (bakejeyner) [reply features/security/src/main/feature/feature.xml:278]: Another good catch 👍

File: features/security/src/main/feature/feature.xml (hunk 6 above)
[the remark this answers is not in the document]
Comment (SmithJosh) [reply features/security/src/main/feature/feature.xml:101]: Huh, that is weird that it couldn't find `org.codice.ddf.platform.util` with that feature installed. platform-email-impl pulls in the security logger service, though, so it does depend on `security-core-services`

[the remark this answers is not in the document]
Comment (bakejeyner) [reply features/security/src/main/feature/feature.xml:101]: Oh I think I know what the issue _could_ have been.

I forgot to add the `security` feature file dependency to the `utilitlies` feature pom. When I did, maven noted a cyclical dependency. Both having a cyclical dependency and not including the `security` feature file as a dependency could have caused the weird behavior I observed.

TLDR: I'm keeping the email bundles in place or else it introduces a cyclical dependency.

**Comment by bakejeyner** [comment]
build now

**Comment by SmithJosh** [comment]
Hero successful. Tested these changes with the Alliance PR linked in the description. For more information see https://github.com/codice/alliance/pull/829#issuecomment-658351148

**Comment by bakejeyner** [comment]
build now

**Comment by bakejeyner** [comment]
build now
